src/grpc_client.cc

```diff
@@ -23,6 +25,48 @@ namespace grpc_labview
 
     //---------------------------------------------------------------------
     //---------------------------------------------------------------------
+    static bool ichar_equals(char a, char b)
+    {
+        return std::tolower(static_cast<unsigned char>(a)) == std::tolower(static_cast<unsigned char>(b));
+    }
+
+    static bool IsLoopbackAddress(const char* address)
+    {
+        std::string hostname(address);
+
+        std::regex hostFromUriRegex(R"(^\w+://([^/?#:]+))");
+        std::smatch match;
+
+        if (std::regex_search(hostname, match, hostFromUriRegex) && match.size() > 1)
+        {
+            hostname = match.str(1); // The host part is in the first capture group
+        }
+        else
+        {
+            size_t lastColonPosition = hostname.rfind(':');
+            if (lastColonPosition != std::string::npos)
+            {
+                hostname = hostname.substr(0, lastColonPosition);
+            }
+        }
+
+        std::string localhost("localhost");
+        bool isLoopback = hostname.size() == localhost.size() && std::equal(hostname.begin(), hostname.end(), localhost.begin(), ichar_equals);
+        if (!isLoopback)
+        {
+            try
+            {
+                asio::ip::address ip_address = asio::ip::make_address(hostname);
+                isLoopback = ip_address.is_loopback();
```

Comment:
Do you really need boost.asio for this?

You can use [inet_pton](https://www.man7.org/linux/man-pages/man3/inet_pton.3.html) to try parsing the address as both IPv4 and IPv6. 
- For IPv4, use `AF_INET` and `struct in_addr`. To check for loopback, I think you can use  something like `(ntohl(x.s_addr) & 0xFF000000) == (INADDR_LOOPBACK & 0xFF000000)` (untested).
- For IPv6, use `AF_INET6` and `struct in6_addr`. To check for loopback, use the [IN6_IS_ADDR_LOOPBACK](https://man7.org/linux/man-pages/man0/netinet_in.h.0p.html) macro.

Reply:
The underlying headers/source ends up using that library so yes I'm sure it can be done. I was trying to avoid reinventing the wheel and learning about the minutiae of that lower level library. Since this is headers only and doesn't require external linkage, I didn't really seem the harm to using it.

BTW, this isn't technically boost.asio. It's just asio. Some of the differences are detailed [here](https://think-async.com/Asio/AsioAndBoostAsio.html). My understanding is that boost.asio is occasionally updated from asio.

Reply:
Okay, plain asio sounds lighter-weight than boost.asio, so I'll close this.

BTW, I was curious whether gRPC has a function for this, and it does, but it's an internal implementation detail and it only checks against 127.0.0.1 / ::1, not other addresses in the loopback subnets.
https://github.com/grpc/grpc/blob/f34b4faca4bbad70ddee456b6e2cfd156a253392/src/core/server/xds_server_config_fetcher.cc#L955